### PR TITLE
Initial debugger UI

### PR DESCRIPTION
**Pull request by samelhusseini**
Todo: 
- [x] Compilation / decompilation of the debugger keyword
- [x] Breakpoint block and field editor

screen shot 2018-04-10 at 3 34 12 pm
screen shot 2018-04-10 at 3 34 08 pm

- [x] Debugger toolbar

screen shot 2018-04-10 at 3 31 03 pm

- [ ] Debugger toggle / entry point (currently the Debug button under the simulator, should go through design review to see what this should look like and how it will interact with tracing)

screen shot 2018-04-10 at 3 30 55 pm

- [ ] Play nice / integrate with tracing (slow mo)
- [x] Debug toolbar can be moved around (top rail)
- [x] Advanced options for JS debugging (step into / out)

screen shot 2018-04-10 at 3 32 06 pm

- [x] Decompilation tests
- [x] Debugging Variables in blocks

screen shot 2018-04-10 at 3 32 55 pm

- [ ] Debug variables in JS
- [x] Debugging is behind the sim debugger flag

### CODE REVIEW

**Comment by pelikhan** [comment]
Comment: i tend to like the toggle on/off better than the current red dot.

File: pxtblocks/blocklyloader.ts [hunk 1]
```diff
@@ -3371,6 +3379,30 @@ namespace pxt.blocks {
...
+                that.setInputsInline(false);
+                that.appendDummyInput('ON_OFF')
+                    .appendField(new Blockly.FieldLabel(lf("breakpoint"), undefined), "DEBUGGER")
+                    .appendField(new pxtblockly.FieldBreakpoint("1", {'type': 'number'}), "ON_OFF");
+
+                setHelpResources(this,
+                    pxtc.TS_DEBUGGER_TYPE,
+                    lf("Debugger statement"),
+                    lf("A debugger statement invokes any available debugging functionality"),
+                    '/blocks/javascript-blocks',
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: /javascript/debugger

File: webapp/src/debugger.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        return Object.keys(variables).length == 0 ? <div /> :
+            <div className="ui segment debugvariables">
+                <div className="ui middle aligned list">
+                    {Object.keys(variables).map(variable =>
+                        <div key={variable} className="item">
+                            <div className="ui label image variable" style={{ backgroundColor: pxt.blocks.getNamespaceColor('variables') }}>
+                                <span className="varname">{variable}</span>
+                                <div className="detail">
+                                    <span className="varval">{variables[variable].value + ' '}</span>
+                                    <span className="previousval">{variables[variable].prevValue ? '(' + variables[variable].prevValue + ')' : ''}</span>
```
Comment (pelikhan) [inline webapp/src/debugger.tsx]: IMO remove variables.

Comment (samelhusseini) [reply webapp/src/debugger.tsx]: What!? I thought that's the primary feature you've been wanting.. I don't understand.

Comment (pelikhan) [reply webapp/src/debugger.tsx]: Sorry wrong comment, how about remove parenthesis?

Comment (samelhusseini) [reply webapp/src/debugger.tsx]: Haha you had me worried for a sec.

File: webapp/src/debugger.tsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        this.nextVariables[name] = value;
+    }
+
+    update() {
+        const variables = this.state.variables;
+        Object.keys(this.nextVariables).forEach(k => {
+            const v = this.nextVariables[k];
+            let sv = '';
+            let type = typeof (v);
+            switch (type) {
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply webapp/src/debugger.tsx]: Done.

**Comment by pelikhan** [comment]
Comments: the slow-mo icon should move in the debugger toolbar and be replaced by a debugger button.

**Comment by samelhusseini** [comment]
Agreed we need better integration with the slow mo icon. It's one of the items on the todo list. We may want to consider merging this as it's independent of those buttons and then we can figure out how we want to integrate the two features together once we've shaped up what we want the overall debugging experience to look like.

**Comment by pelikhan** [comment]
Slow Mo is fundamentally a debugger feature so I think it should be lumped in. For example, seeing variables values during slow mo can be very useful or hitting pause, etc...

________________________________

**Comment by samelhusseini** [comment]
Sure. I was just pushing for small incremental changes rather than big pushes, but happy to wait on this PR until we make a decision around slow mo.

**Comment by pelikhan** [comment]
You can push it in. You could ad a slow mo button on the toolbar meanwhile.

(I wasn’t able to get the variable table visible)

________________________________

**Pull request merged**